### PR TITLE
Add new_count_waker

### LINKED ISSUE #1384: Change to WakeCounter API proposal
I would like to propose a breaking change in API for the `WakerCounter` type in the futures-test crate. The current API has the counter and `LocalWaker` implementation in the same `WakerCounter` type, which can be limiting. I propose an API that splits it into two types. Since it's not a lot of code I have a working implementation below.

```rust
/// Create a new `LocalWaker` that counts the number of times it's awoken.
pub fn new_count_waker() -> (LocalWaker, AwokenCount) {
    let inner = Arc::new(WakerInner { count: AtomicUsize::new(0) });
    (local_waker_from_nonlocal(inner.clone()), AwokenCount { inner })
}

/// Number of times the waker was awoken.
///
/// See [`new_count_waker`].
pub struct AwokenCount {
    inner: Arc<WakerInner>,
}

impl AwokenCount {
    /// Get the number of times the waker was awoken.
    pub fn get(&self) -> usize {
        self.inner.count.load(Ordering::SeqCst)
    }
}

struct WakerInner {
    count: AtomicUsize,
}

impl Wake for WakerInner {
    fn wake(arc_self: &Arc<Self>) {
        let _ = arc_self.count.fetch_add(1, Ordering::SeqCst);
    }
}
```

And the usage.

```rust
let (waker, count) = new_count_waker();

// Polling is easy since we just have a `LocalWaker`.
let future = // ...
future.poll(&waker);

// And elsewhere we can check the wake count.
assert_eq!(count.get(), 1);
```

### PR DESCRIPTION
**Pull request by Thomasdezeeuw**
This is an alternative API to WakeCounter, see #1384.

I didn't touch the old API, but it might be a bit confusing that there are two different API's to do the same thing. I'll leave it up to someone else to decide whether to deprecate/remove/leave it.

Closes #1384.

/cc @Nemo157

### CODE REVIEW

**Comment by Thomasdezeeuw** [comment]
Some more notes:

The atomic ordering might be overkill, but since it's a test API I don't think the performance impact matters too much. Also I didn't add a test, but the example tests the entire API already.

**Comment by Thomasdezeeuw** [comment]
I think the Travis failure is unrelated, from the log:

```log
error[E0463]: can't find crate for `std`
  |
  = note: the `thumbv6m-none-eabi` target may not be installed
```

Other architectures are fine.

**Comment by Nemo157** [comment]
Yeah, it’s https://github.com/rust-lang-nursery/futures-rs/issues/1396, I would fix it but I don’t really know if there’s a way to do so that doesn’t break other stuff.

**Comment by Thomasdezeeuw** [comment]
I've force pushed a rebase on master, the CI should pass now as the failure was fixed in #1407.

File: futures-test/src/task/wake_counter.rs [hunk 1]
```diff
@@ -68,3 +68,52 @@ impl Wake for Inner {
...
+
+/// Number of times the waker was awoken.
+///
+/// See [`new_count_waker`] for usage.
+#[derive(Debug)]
+pub struct AwokenCount {
+    inner: Arc<WakerInner>,
+}
+
+impl PartialEq<usize> for AwokenCount {
```
Comment (cramertj) [inline futures-test/src/task/wake_counter.rs]: I think `Deref<Target = usize>` would also be handy.

**Comment by cramertj** [comment]
Seems good to me! I'd prefer we replaced the old one at the same time, though, or at least marked it as deprecated and switched our internal usage. No reason to let tech debt sit around on an alpha version of the library.

**Comment by Thomasdezeeuw** [comment]
@cramertj  I'm going to address the comments now, to respond to your comment in the other issue.

**Comment by cramertj** [comment]
sg, thanks!

File: futures-test/src/task/wake_counter.rs (hunk 1 above)
Comment (Thomasdezeeuw) [reply futures-test/src/task/wake_counter.rs]: How can I do that? I don't think `AtomicUsize` has any methods that allow you to get a reference to the inner usize. I tried the following

```rust
impl Deref for AwokenCount {
    type Target = usize;

    fn deref(&self) -> &usize {
        &self.inner.count.load(Ordering::SeqCst)
    }
}
```

But that gives the `cannot return reference to temporary value` error, which makes sense.

I could add back the `get(&self) -> usize` method

**Comment by Thomasdezeeuw** [comment]
Pushed 2 commits to changes the tests and remove the old `WakeCounter`. Should I implement the `get(&self) -> usize` method?

**Comment by Thomasdezeeuw** [comment]
@cramertj I think the test will pass (they work on my machine ™), if your OK with the API then I'm going to bed now.

**Comment by cramertj** [comment]
Yup, looks good, thanks for the changes!

**Pull request merged**